### PR TITLE
Rework callbacks to remove edge cases

### PR DESCRIPTION
**Pull request by anujdas**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,30 +1,47 @@
...
-    super(`Thrift pool connection acquisition timeout: ${message}`, metadata);
-    this.name = 'AcquisitionTimeoutError';
-    Error.captureStackTrace(this, AcquisitionTimeoutError);
-  }
+/**
+ * Wraps a connection in event callbacks suitable for Promise use
+ *
+ * @param {Connection} connection A Thrift connection
+ * @param {function} reject A Promise's reject function
+ *
```
Comment (MatthewHerbst) [inline src/index.js:12]: Nit: newline here not needed

File: src/index.js [hunk 2]
```diff
@@ -1,30 +1,47 @@
...
+  };
+  const onClose = () => {
+    connection.alive = false;
+    reject(new ConnectionClosedError());
+  };
+  const onError = (error) => {
+    connection.alive = false;
+    reject(error);
+  };
+  connection.on('timeout', onTimeout).on('close', onClose).on('error', onError);
```
Comment (MatthewHerbst) [inline src/index.js:28]: Nit:

```
connection
  .on(...)
  .on(...)
  .on(...)
```

File: src/index.js [hunk 3]
```diff
@@ -1,30 +1,47 @@
...
-  }
+/**
+ * Removes callbacks attached to a connection by attachCallbacks
+ *
+ * @param {Connection} connection A Thrift connection
+ * @param {Object} callbacks The returned { onTimeout, onClose, onError } from attachCallbacks
+ */
+const detachCallbacks = (connection, { onTimeout, onClose, onError }) => {
+  connection
+    .removeListener('timeout', onTimeout)
```
Comment (MatthewHerbst) [inline src/index.js:42]: Does `removeListener` promise to never throw?

**Review by MatthewHerbst — APPROVED** [review]

File: src/index.js (hunk 2 above)
Comment (anujdas) [reply src/index.js:28]: ugh I really hate all the extraneous newlines

File: src/index.js (hunk 3 above)
Comment (anujdas) [reply src/index.js:42]: So it seems, based on https://nodejs.org/api/events.html#events_emitter_removelistener_eventname_listener

File: src/index.js (hunk 1 above)
Comment (anujdas) [reply src/index.js:12]: True

**Comment by anujdas** [comment]
Pulling in changes directly

**Pull request closed without merging**